### DISCUSSION TITLE
can't call computeOffDiagJacobianScalar in a NodalKernel?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#23267] [Q&A General] [answered]
A `NodalKernel` has a `computeOffDiagJacobianScalar` method accoring to [doxygen](https://mooseframework.inl.gov/docs/doxygen/moose/classNodalKernel.html), but it doesn't seem to be getting called in my nodal kernel. I have scalar variables in my system, and they are coupled in my constructor (`coupledScalar`). Is there something more I need to do? FWIW, my on-diagonal Jacobian method is being called. I also threw in a `computeQpOffDiagJacobian` method that does nothing, but that doesn't seem be getting called either, which is odd because I have 2 nonlinear displacement variables in my system (it is a 2D mechanics problem).

**Comment by hugary1995** [comment, marked as answer]
It is off diagonal w.r.t. other scalar variables.

**Reply by jessecarterMOOSE** [reply]
Oh, didn't realize that. Well, I have 3 scalar variables in my system so I would still expect it to be called!

Then I guess my question should be - what is the proper method for computing the Jacobian contribution for nonlinear variables with respect to scalar variables and vice versa?

**Reply by ttruster** [reply]
We had some discussion about a similar issue for Mortar Constraints over here https://github.com/idaholab/moose/discussions/22020. Basically, the `computeOffDiagJacobianScalar` method is only implemented within the assembly loops for `Kernels` and `IntegratedBC` within MOOSE; none of the other `ResidualObjects` address that method. I have an open pull request at https://github.com/idaholab/moose/pull/22376 to implement interfaces for coupling scalar variables to `Kernels` and `MortarConstraints` more effectively. My next plan is for `InterfaceKernels` and maybe `IntegratedBC`, although it won't be for a while. You can look there for some inspiration and discuss sometime if you want. The basic idea is to add a focus scalar variable to the `Kernel` object and then do assembly of that scalar within the `computeResidual` and `computeJacobian` contributions to the field variables.

**Reply by jessecarterMOOSE** [reply]
FWIW, what I ended up doing was overriding `computeJacobian` in both my nodal kernel and scalar kernel and setting Jacobian entries manually with `_assembly.cacheJacobian(row, col, value,...)`, where the row and column come from the DoF associated with the kernel's variable (or coupled variable) and node via the `dof_number` method of the `Node` class.
